code/obj/item/cigarette.dm

```diff
@@ -569,39 +576,55 @@
 
 /obj/item/cigpacket/proc/update_icon()
 	src.overlays = null
-	if (src.cigcount <= 0)
+	if (length(src.cigs) == 0)
 		src.icon_state = "[src.package_style]0"
 		src.desc = "There aren't any cigs left, shit!"
 	else
 		src.icon_state = "[src.package_style]o"
-		src.overlays += "cig[src.cigcount]"
+		src.overlays += "cig[length(src.cigs)]"
 	return
 
 /obj/item/cigpacket/attack_hand(mob/user as mob)
 	if (user.find_in_hand(src))//r_hand == src || user.l_hand == src)
-		if (src.cigcount == 0)
+		if (length(src.cigs) == 0)
 			user.show_text("You're out of cigs, shit! How you gonna get through the rest of the day?", "red")
 			return
 		else
-			var/obj/item/clothing/mask/cigarette/W = new src.cigtype(user)
+			var/obj/item/clothing/mask/cigarette/W = src.cigs[1]
 			user.put_in_hand_or_drop(W)
-			if (src.cigcount != -1)
-				src.cigcount--
+			src.cigs -= W
```

Comment:
Generally it's fast to remove from the end of a list and slow to remove from the beginning (because all other elements need to be shifted one position left). But this is a tiny list so it doesn't matter.

Reply:
Since it's small enough to not matter, I'll leave it like this. This implementation has the advantage of letting you "shuffle" through a pack of different cigarettes, since it's FILO, whereas a FIFO implementation, while faster, would be less convenient for reordering a mixed pack. (since you'd just be adding and removing the same cig each time you click)